observer/obs_conf.go

```diff
@@ -133,5 +134,9 @@ func (c *ObsConf) MakeObserver() (*Observer, error) {
 	if err != nil {
 		return nil, err
 	}
+	// Register prober-specific metrics as applicable
+	for _, collector := range probers.ProberCollectors {
+		metrics.MustRegister(collector)
+	}
```

Comment:
It turns out that this doesn't actually impact the ability for multiple probers to publish their metrics. If you construct two separate but equivalent `prometheus.Collector` objects (same options and labels), they can both be used to make observations to the metric they define, regardless of which one was passed to `MustRegister()`. Granted, you'll still be using more memory than strictly necessary to hold the duplicate objects

Reply:
I don't believe this is true. Trying running this program, and you will see that the `Gather()` call (which is what is [used behind the scenes](https://github.com/letsencrypt/boulder/blob/6a9bb399f7f7a8405e2af47fc64d7469331a47ed/vendor/github.com/prometheus/client_golang/prometheus/promhttp/http.go#L126) by the [http endpoint](https://github.com/letsencrypt/boulder/blob/6a9bb399f7f7a8405e2af47fc64d7469331a47ed/cmd/shell.go#L218) scraped by our metrics gathering system) only sees the updates to the registered metric, not to the unregistered one.
```go
package main

import (
	"errors"
	"fmt"
	"os"

	"github.com/prometheus/client_golang/prometheus"
)

func helper() error {
	registry := prometheus.NewRegistry()

	// Create and register a metric.
	counter := prometheus.NewCounterVec(prometheus.CounterOpts{
		Name: "counter",
		Help: "Test counter metric",
	}, []string{"registered"})
	err := registry.Register(counter)
	if err != nil {
		return fmt.Errorf("registering metric: %w", err)
	}

	// Create (but fail to register) an identical metric.
	dupCounter := prometheus.NewCounterVec(prometheus.CounterOpts{
		Name: "counter",
		Help: "Test counter metric",
	}, []string{"registered"})
	err = registry.Register(dupCounter)
	if err == nil {
		return errors.New("registering duplicate metric: should have resulted in error but didn't")
	}

	// Increment both the registered and unregistered versions of the metric.
	counter.WithLabelValues("true").Inc()
	dupCounter.WithLabelValues("false").Inc()

	// Gather all metrics that the registry is aware of.
	metricFamilies, err := registry.Gather()
	if err != nil {
		return fmt.Errorf("gathering metrics: %w", err)
	}

	// Let's see what we got
	for i, mf := range metricFamilies {
		fmt.Printf("Family %d: %s\n", i, mf.Name)
		for j, m := range mf.Metric {
			fmt.Printf("  Metric %d: %s\n", j, *m.Summary)
		}
	}
}

func main() {
	err := helper()
	if err != nil {
		fmt.Println(err)
		os.Exit(1)
	}
}
```

Reply:
We should point out that, yes, a whole new object will have a different prometheus ID, but a copy of the previous object (which is what I'm suggesting in my comments below) will work just fine:

```go
package main

import (
	"fmt"
	"os"

	"github.com/prometheus/client_golang/prometheus"
)

func helper() error {
	registry := prometheus.NewRegistry()

	// Create and register a metric.
	counter := prometheus.NewCounterVec(
		prometheus.CounterOpts{
			Name: "counter",
			Help: "Test counter metric",
		}, []string{"registered"})
	err := registry.Register(counter)
	if err != nil {
		return fmt.Errorf("registering metric: %w", err)
	}

	// Make a copy of the metric.
	counterCopy := *counter

	// Increment both the registered and unregistered versions of the metric.
	counter.WithLabelValues("true").Inc()
	counterCopy.WithLabelValues("false").Inc()

	// Gather all metrics that the registry is aware of.
	metricFamilies, err := registry.Gather()
	if err != nil {
		return fmt.Errorf("gathering metrics: %w", err)
	}

	// Let's see what we got
	for i, mf := range metricFamilies {
		fmt.Printf("Family %d: %s\n", i, *mf.Name)
		for j, m := range mf.Metric {
			fmt.Printf("  Metric %d:\n", j)
			for _, l := range m.Label {
				fmt.Printf("    %s = %s\n", *l.Name, *l.Value)
			}
			fmt.Printf("    %0.0f\n", *m.Counter.Value)
		}
	}
	return nil
}

func main() {
	err := helper()
	if err != nil {
		fmt.Println(err)
		os.Exit(1)
	}
}
```

When I run this, the output is:
```
$ go run ./main.go
Family 0: counter
  Metric 0:
    registered = false
    1
  Metric 1:
    registered = true
    1
```

Reply:
I dug into this a bit more just to make sure I have a handle on why this is the case. When a `prometheus.Collector` is registered with a prometheus registry, it keeps a copy of that collector and watches it's inner metrics channel (type `chan`) for metrics to process. When you make a copy of an object which contains a channel, that channel is still a reference. However, when you create a whole new instance of a collector, it will have a completely different channel that it will be sending metrics to which the prometheus Registry will be completely unaware of.